core/include/edm/container.hpp

```diff
@@ -51,9 +84,163 @@ namespace traccc {
+            header_vector && hv,
+            item_vector && iv
+        )
+            : headers(hv)
+            , items(iv)
+        {
+            if (headers.size() != items.size()) {
+                throw std::logic_error("Header and item length not equal.");
+            }
+        }
+
+        /**
+         * @brief Argument-forwarding constructor.
+         *
+         * This constructor forwards its arguments to both internal vectors.
+         * Obviously, this requires the arguments to be quite general, but this
+         * constructor can be useful when initializing a container for a
+         * specific vecmem memory resource.
+         */
+        template<class... args>
+        container(
+            args&&... a
+        )
+            : headers(a...)
+            , items(a...)
+        {}
+
+        /**
+         * @brief Bounds-checking mutable element accessor.
+         */
+        element_view
+        at(
+            const size_type & i
+        ) {
+            if (i >= size()) {
+                throw std::out_of_range("Index out of range.");
+            }
+
+            return operator[](i);
+        }
+
+        /**
+         * @brief Bounds-checking immutable element accessor.
+         */
+        const element_view
+        at(
+            const size_type & i
+        ) const {
+            if (i >= size()) {
+                throw std::out_of_range("Index out of range.");
+            }
+
+            return operator[](i);
+        }
+
+        /**
+         * @brief Mutable element accessor.
+         */
+        element_view
+        operator[](
+            const size_type & i
+        ) {
+            return {headers.at(i), items.at(i)};
+        }
+
+        /**
+         * @brief Immutable element accessor.
+         */
+        const element_view
+        operator[](
+            const size_type & i
+        ) const {
+            return {headers.at(i), items.at(i)};
+        }
+
+        /**
+         * @brief Return the size of the container.
+         *
+         * In principle, the size of the two internal vectors should always be
+         * equal, but we can assert this at runtime for debug builds.
+         */
+        size_type
+        size(
+            void
+        ) const {
+            #ifndef NDEBUG
+            if (headers.size() != items.size()) {
+                throw std::logic_error("Header and item length not equal.");
+            }
+            #endif
+
+            return headers.size();
+        }
+
+        /**
+         * @brief Reserve space in both vectors.
+         */
+        void
+        reserve(
+            const size_type & s
+        ) {
+            headers.reserve(s);
+            items.reserve(s);
+        }
+
+        /**
+         * @brief Push a header and a vector into the container.
+         */
+        template<typename h_prime, typename v_prime>
+        void
+        push_back(
+            h_prime && new_header,
+            v_prime && new_items
```

Comment:
There should be a "copy version" of this function as well, if we want to have this functionality.

Reply:
I disagree here, implicitly copying data here is a performance anti-pattern. If the user does want to copy, they can do so explicitly without our code having to care about it.